### PR TITLE
New Extension: pip_compile - a robust successor to the pip_compile_button extension

### PR DESCRIPTION
**Pull request by TimothyLoyer**
# What this PR is for
Problems with `pip_compile_button`:
- K8s deployment must be in an error free state
- Requirements could not be compiled if requirements were in a broken state *
- File changes in k8s deployment can interrupt pip-compile

These problems are all addressed by the new `pip_compile` extension, because it runs outside of Kubernetes using `docker run`. It can also be configured to build more quickly by targeting an earlier build stage before any `pip-tools` commands have run in the image.

* Broken requirements can be rebuilt so long as the build stage target is _before_ `pip-compile` is run (and the target stage itself does not call it).

Bonus: The `pip-compile` button is automagically added to _all_ Tilt workloads which use same image!

# Notes to Reviewer
To test this, you will need to clone this repo and point to its filepath rather than github for our custom tilt extensions:

```starlark
v1alpha1.extension_repo(
    name = "de-tilt",
    url = "file:///home/.../tilt-extensions",
)
```

Then register the new extension (`pip_compile` rather than `pip_compile_button`):

```starlark
v1alpha1.extension(
    name = "pip_compile",
    repo_name = "de-tilt",
    repo_path = "pip_compile",
)
```

...and load it as usual (note: the function is still named `pip_compile_button`)

```starlark
load("ext://pip_compile", "pip_compile_button")
```

Finally, invoke the extension to create the `pip-compile` buttons:

```starlark
pip_compile_button(
    "image-ref",
    target = "build",  # optional: target an earlier stage
    reqs_path = "./requirements/"  # optional: local path to directory containing requirements files
    compile_args = ["--upgrade-package", "black"],  # optional: change default arguments sent to `pip-compile`
)
```

### CODE REVIEW

File: pip_compile/scripts/pip-compile-container.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+echo "Please wait: Building Docker image..."
+docker build --quiet \
+  --tag "$BUILD_TAG" \
+  --target "$IMAGE_TARGET" \
+  --cache-from "$IMAGE_REF" \
+  --file "$DOCKERFILE" "$IMAGE_CONTEXT"
+
+# Compile pip requirements.
+echo "Please wait: Compiling pip requirements with pip-tools..."
+docker run -i --rm \
```
Comment (noelleleigh) [inline pip_compile/scripts/pip-compile-container.sh:29]: If this fails for whatever reason, should the user get a nice message telling them that the `pip-compile` operation wasn't successful? Right now if pip dies, a stack trace is printed and then nothing else.

Same for the `docker build` step.

File: pip_compile/scripts/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+cd "$BUILD_PATH"
+
+# Copy source files to build directory
+cp $MOUNT_PATH/*.in ./
+cp $MOUNT_PATH/*.txt ./
+
+# Compile requirements
+for filepath in "$BUILD_PATH"/*.in; do
+  requirements=$(basename "$filepath")
+  pip-compile --generate-hashes --allow-unsafe "$requirements"
```
Comment (noelleleigh) [inline pip_compile/scripts/entrypoint.sh:24]: A use-case that the current pip_compile_button extension supports is adding additional arguments to pip-compile. For example, if I want to update a sub-dependency, I can edit my Tiltfile to pass the [`--upgrade-package` argument](https://github.com/jazzband/pip-tools#updating-requirements) temporarily, then remove it when it's done.

File: pip_compile/scripts/pip-compile-container.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+# Script Paths
+LOCAL_SCRIPT_PATH=$(dirname "$0")
+REQ_SCRIPT_PATH="/.pip-requirements/scripts/"
+
+# Build image.
+echo "Please wait: Building Docker image..."
+docker build --quiet \
+  --tag "$BUILD_TAG" \
+  --target "$IMAGE_TARGET" \
```
Comment (noelleleigh) [inline pip_compile/scripts/pip-compile-container.sh:23]: If the Dockerfile doesn't have a target between pip-tools being installed and the full requirements being installed, a faulty requirements.txt file can cause this build process to fail as well:

```
Running cmd: sh /Users/noelle/dev/DirectEmployers/tilt-extensions/pip_compile/scripts/pip-compile-container.sh postajob /Users/noelle/dev/DirectEmployers/postajob Dockerfile backend .
Please wait: Building Docker image...
Dockerfile:58
--------------------
  57 |     COPY --chown=django requirements.txt ./
  58 | >>> RUN --mount=type=cache,target=$CACHE_DIR,uid=1000,gid=1000 \
  59 | >>>     pip install -r requirements.txt
  60 |     COPY --chown=django requirements.in ./
--------------------
ERROR: failed to solve: process "/bin/sh -c pip install -r requirements.txt" did not complete successfully: exit code: 1

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/7y31u6b0cf2r79mdbltpt1u6v
```

This means the "Requirements could not be compiled if requirements were in a broken state" problem isn't solved by this.

Comment (TimothyLoyer) [reply pip_compile/scripts/pip-compile-container.sh:23]: I added an asterisk to that bullet point in the PR.

**Review by noelleleigh — COMMENTED** [review]
Can you update your PR description with the changed API for pip_compile_button?

File: pip_compile/scripts/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env sh
+set -eux
+
+BUILD_PATH="/.pip-requirements/build"
+MOUNT_PATH="/.pip-requirements/mount"
+
+diffcp() {
+  # Copy files only when their contents have changed (prevents Tilt reload)
+  cmp -s "$1" "$2" || cp "$1" "$2" && echo "No changes made to $1."
```
[the remark this answers is not in the document]
Comment (TimothyLoyer) [reply pip_compile/scripts/entrypoint.sh:9]: Good idea - this change will be pushed soon.

File: pip_compile/Tiltfile [hunk 5]
```diff
@@ -0,0 +1,45 @@
+load("ext://uibutton", "cmd_button")
+
+
+SCRIPTS = os.path.join(os.getcwd(), "scripts")
+DEFAULT_COMPILE_ARGS = "--allow-unsafe --generate-hashes"
+
+
+def pip_compile_button(
```
[the remark this answers is not in the document]
Comment (TimothyLoyer) [reply pip_compile/Tiltfile:8]: Fixed (and overhauled to use python more)

**Comment by TimothyLoyer** [comment]
I've overhauled this to use Python. I did this because I was finding myself passing around variables both repetitively and in ways which were quite ugly (i.e. the contents of a Dockerfile, made necessary by the lack of ability to know where the Dockerfile was located relative to the context path).

The good news is... it has worked quite well in my testing.

File: pip_compile/Tiltfile [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+DEFAULT_COMPILE_ARGS = ["--allow-unsafe", "--generate-hashes"]
+
+
+def pip_compile_button(
+    image,
+    target = "",
+    reqs_path = ".",
+    compile_args = DEFAULT_COMPILE_ARGS,
+):
+    """Add `pip-compile` button to all Tilt resources using `image`.
```
Comment (noelleleigh) [inline pip_compile/Tiltfile:14]: My editor wants the function arguments to be documented in the docstring, and I agree.

image

File: pip_compile/scripts/controller.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+def tilt_resource_template(kind: str) -> dict:
+    return {
+        "apiVersion": "tilt.dev/v1alpha1",
+        "kind": kind,
+        "metadata": {},
+        "spec": {},
+    }
+
+
+def generate_tilt_json(resource: str, context: str, arguments: list[str]) -> str:
```
Comment (noelleleigh) [inline pip_compile/scripts/controller.py:100]: Could you document this function to explain what it does?

File: pip_compile/Tiltfile [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    reqs_path = os.path.realpath(reqs_path)
+
+    # Start a controller to watch for image changes and update related resources.
+    cmd = ["python", SCRIPT, image, "--target", target, "--reqs-path", reqs_path]
+    for arg in compile_args:
+        cmd.append("--compile-args=%s" % arg)
+
+    controller_resource = "pip-compile-controller:%s" % image
+    local_resource(
+        name = controller_resource,
```
Comment (noelleleigh) [inline pip_compile/Tiltfile:29]: An image name could include `/`, but Tilt doesn't like that:

```
Loading Tiltfile at: /Users/noelle/dev/DirectEmployers/postajob/Tiltfile
Traceback (most recent call last):
  /Users/noelle/dev/DirectEmployers/postajob/Tiltfile:65:19: in <toplevel>
  /Users/noelle/dev/DirectEmployers/tilt-extensions/pip_compile/Tiltfile:28:19: in pip_compile_button
Error in local_resource: local_resource: for parameter "name": invalid value "pip-compile-controller:postajob/foo/bar": may not contain '/'
```

Can you tweak `controller_resource` to make it Tilt-compatible in that situation?

File: pip_compile/README.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+load("ext://pip_compile", "pip_compile_button")
+
+pip_compile_button(    
+    # Name of the Docker image (same as the first argument of `docker_build`).
+    "image-name",
+    
+    # Optional: Name of build stage to target (same as the `target` argument of `docker_build`).
+    target = "build",
+    
+    # Path to the directory which contains the requirements .in and .txt files.
```
Comment (noelleleigh) [inline pip_compile/README.md:21]:
```suggestion
    # Optional: Path to the directory which contains the requirements .in and .txt files (Default: `.`).
```

File: pip_compile/Tiltfile [hunk 10]
```diff
@@ -0,0 +1,39 @@
+load("ext://uibutton", "cmd_button")
+
+
+SCRIPT = os.path.join(os.getcwd(), "scripts", "controller.py")
+DEFAULT_COMPILE_ARGS = ["--allow-unsafe", "--generate-hashes"]
+
+
+def pip_compile_button(
```
Comment (noelleleigh) [inline pip_compile/Tiltfile:8]: When I try to start the postajob project with 

```
pip_compile_button("postajob")
```

in the Tiltfile, the app resources never come up and the controller seems to hang:

image

File: pip_compile/scripts/compiler.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+def build_image(image, tilt_resource, target):
+    dockerimages_json = subprocess.run(
+        ["tilt", "get", "dockerimages", f"{tilt_resource}:{image}", "-o", "json"],
+        capture_output=True,
+        text=True,
+    )
+    dockerimages = json.loads(dockerimages_json.stdout)
+    spec = dockerimages["spec"]
+
+    # with tempfile.NamedTemporaryFile(mode="w+") as dockerfile:
```
Comment (noelleleigh) [inline pip_compile/scripts/compiler.py:25]: Commented code

File: pip_compile/scripts/compiler.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
+import argparse
+import json
+import subprocess
+import tempfile
```
Comment (noelleleigh) [inline pip_compile/scripts/compiler.py:4]: Unused import

File: pip_compile/scripts/compiler.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    ]
+    process = subprocess.run(
+        docker_build,
+        input=spec["dockerfileContents"].encode("utf-8"),
+    )
+
+    if process.returncode != 0:
+        print("Error: Docker image failed to build")
+        print(process.stdout, process.stderr)
+        exit(1)
```
Comment (noelleleigh) [inline pip_compile/scripts/compiler.py:50]: Could error checking be handled by passing `check=True` to [`subprocess.run()`](https://docs.python.org/3/library/subprocess.html#subprocess.run)?

File: pip_compile/scripts/compiler.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+BUILD_TAG = "build"
+
+SCRIPTS = Path(__file__).parent
+
+# Container Paths
+CONTAINER_REQUIREMENTS_PATH = "/.pip-requirements/mount/"
+CONTAINER_SCRIPTS_PATH = "/.pip-requirements/scripts/"
+
+
+def build_image(image, tilt_resource, target):
```
Comment (noelleleigh) [inline pip_compile/scripts/compiler.py:16]: Can you add type hints and docstring to this function?

File: pip_compile/scripts/compiler.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+        input=spec["dockerfileContents"].encode("utf-8"),
+    )
+
+    if process.returncode != 0:
+        print("Error: Docker image failed to build")
+        print(process.stdout, process.stderr)
+        exit(1)
+
+
+def run_container(image, compile_args, local_req_path):
```
Comment (noelleleigh) [inline pip_compile/scripts/compiler.py:53]: Can you add type hints and docstring to this function?

File: pip_compile/scripts/compiler.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        f"--volume={local_req_path}:{CONTAINER_REQUIREMENTS_PATH}:rw",
+        f"--volume={SCRIPTS}:{CONTAINER_SCRIPTS_PATH}:ro",
+        f"--entrypoint={CONTAINER_SCRIPTS_PATH}/entrypoint.sh",
+        f"{image}:{BUILD_TAG}",
+    ]
+    docker_run.extend(compile_args)
+    process = subprocess.run(docker_run)
+    if process.returncode != 0:
+        print("Error: Pip requirements failed to compile")
+        exit(2)
```
Comment (noelleleigh) [inline pip_compile/scripts/compiler.py:70]: Could error checking be handled by passing `check=True` to [`subprocess.run()`](https://docs.python.org/3/library/subprocess.html#subprocess.run)?

File: pip_compile/scripts/compiler.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    compile_args: list[str],
+):
+    """Build and run image to build pip requirements with pip-tools."""
+    build_image(image, resource, target)
+
+    # Compile pip requirements.
+    print("Please wait: Compiling pip requirements with pip-tools...")
+    run_container(image, compile_args, local_req_path)
+
+    print("Done: Pip requirements have been compiled!")
```
Comment (noelleleigh) [inline pip_compile/scripts/compiler.py:87]:
```suggestion
    print("Done: pip requirements have been compiled!")
```

File: pip_compile/Tiltfile (hunk 6 above)
Comment (TimothyLoyer) [reply pip_compile/Tiltfile:14]: This can be done :)

File: pip_compile/Tiltfile (hunk 8 above)
Comment (TimothyLoyer) [reply pip_compile/Tiltfile:29]: Good find!!

File: pip_compile/scripts/compiler.py (hunk 13 above)
Comment (TimothyLoyer) [reply pip_compile/scripts/compiler.py:50]: Great idea!

File: pip_compile/scripts/compiler.py (hunk 15 above)
Comment (TimothyLoyer) [reply pip_compile/scripts/compiler.py:53]: Yes, not sure how I missed this one!

**Review by noelleleigh — APPROVED** [review]
Great work on this!

**Pull request merged**